File: src/main/java/komu/blunt/core/CoreApplicationExpression.java
package komu.blunt.core;

import komu.blunt.asm.*;

import static com.google.common.base.Preconditions.checkNotNull;

public final class CoreApplicationExpression extends CoreExpression {
    
    private final CoreExpression func;
    private final CoreExpression arg;

    public CoreApplicationExpression(CoreExpression func, CoreExpression arg) {
        this.func = checkNotNull(func);
        this.arg = checkNotNull(arg);
    }

    @Override
    public Instructions assemble(Assembler asm, Register target, Linkage linkage) {
        
        Instructions instructions = new Instructions();
        
        instructions.append(func.assemble(asm, Register.PROCEDURE, Linkage.NEXT));
        instructions.append(arg.assemble(asm, Register.ARG, Linkage.NEXT).preserving(Register.PROCEDURE));

        if (linkage == Linkage.RETURN && target == Register.VAL) {
            instructions.applyTail();

        } else {
            Label afterCall = asm.newLabel("afterCall");

            // TODO: make pushing env the responsibility of called procedure
            if (linkage != Linkage.RETURN) instructions.pushRegister(Register.ENV);

            // TODO: use label from linkage if possible (depends on callee saving env)
            instructions.pushLabel(afterCall);
            instructions.apply();
            instructions.label(afterCall);
            if (linkage != Linkage.RETURN) instructions.popRegister(Register.ENV);

            if (target != Register.VAL)
                instructions.copy(target, Register.VAL);

            instructions.finishWithLinkage(linkage);
        }


        return instructions;
    }

    @Override
    public CoreExpression simplify() {
        // TODO: simplify application of lambda
        return new CoreApplicationExpression(func.simplify(), arg.simplify());
    }

    @Override
    public String toString() {
        return "(" + func + " " + arg + ")";
    }
}

File: src/main/kotlin/komu/blunt/core/CoreApplicationExpression.kt
package komu.blunt.core

import komu.blunt.asm.*
import komu.blunt.asm.Register.*

class CoreApplicationExpression(private val func: CoreExpression,
                                private val arg: CoreExpression) : CoreExpression() {

    override fun assemble(asm: Assembler, target: Register, linkage: Linkage) =
        instructions {
            instructionsOf(func.assemble(asm, PROCEDURE))

            preserving(PROCEDURE) {
                instructionsOf(arg.assemble(asm, ARG))
            }

            if (linkage == Linkage.Return && target == VAL) {
                applyTail()

            } else {
                val afterCall = asm.newLabel("afterCall");

                // TODO: make pushing env the responsibility of called procedure
                if (linkage != Linkage.Return) pushRegister(ENV)

                // TODO: use label from linkage if possible (depends on callee saving env)
                pushLabel(afterCall)
                apply()
                label(afterCall)
                if (linkage != Linkage.Return) popRegister(ENV)

                if (target != VAL)
                    copy(target, VAL)

                finishWithLinkage(linkage)
            }
        }

    override fun simplify() =
        // TODO: simplify application of lambda
        CoreApplicationExpression(func.simplify(), arg.simplify())

    override fun toString() = "($func $arg)"
}
